File: Backend/LeerIA-backend/app/api/v1/documents.py
from fastapi import APIRouter, HTTPException, UploadFile, File, Form
from pathlib import Path
from uuid import UUID
import shutil

from app.database.database_call import supabase
from app.models.entities import DocumentCreate


router = APIRouter()


@router.post("/")
def create_document(document: DocumentCreate):
    data = document.model_dump(mode="json")

    try:
        subject_response = (
            supabase
            .table("subjects")
            .select("*")
            .eq("id", data["subject_id"])
            .execute()
        )

        if not subject_response.data:
            raise HTTPException(
                status_code=404,
                detail="La materia asociada no existe"
            )

        response = (
            supabase
            .table("documents")
            .insert(data)
            .execute()
        )

        return {
            "message": "Documento creado correctamente",
            "data": response.data[0]
        }

    except HTTPException:
        raise

    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail=f"Al crear documento. Error: {str(error)}"
        )





BASE_DIR = Path(__file__).resolve().parents[3]
UPLOAD_DIR = BASE_DIR / "storage" / "uploads"

ALLOWED_EXTENSIONS = {".pdf", ".docx", ".pptx", ".txt"}


@router.post("/upload")
async def upload_document( subject_id: UUID = Form(...), file: UploadFile = File(...)):

    try:
       
        subject_response = (
            supabase
            .table("subjects")
            .select("*")
            .eq("id", str(subject_id))
            .execute()
        )

        if not subject_response.data:
            raise HTTPException(
                status_code=404,
                detail="La materia asociada no existe"
            )

        
        if not file.filename:
            raise HTTPException(
                status_code=400,
                detail="El archivo no tiene nombre"
            )

        file_extension = Path(file.filename).suffix.lower()

        if file_extension not in ALLOWED_EXTENSIONS:
            raise HTTPException(
                status_code=400,
                detail=f"Tipo de archivo no permitido: {file_extension}"
            )

        
        subject_upload_dir = UPLOAD_DIR / str(subject_id)
        subject_upload_dir.mkdir(parents=True, exist_ok=True)

        
        file_path = subject_upload_dir / file.filename

        
        with file_path.open("wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        
        document_data = {
            "subject_id": str(subject_id),
            "file_name": file.filename,
            "file_type": file.content_type,
            "storage_path": str(file_path),
            "status": "uploaded",
        }

        document_response = (
            supabase
            .table("documents")
            .insert(document_data)
            .execute()
        )

        return {
            "message": "Documento subido correctamente",
            "data": document_response.data[0]
        }

    except HTTPException:
        raise

    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail=f"Error al subir documento: {str(error)}"
        )

    finally:
        await file.close()

File: Frontend/LeerIA-frontend/src/shared/api/documents.ts
const API_URL = import.meta.env.VITE_API_URL;

export type ApiDocument = {
  id: string;
  subject_id: string;
  file_name: string;
  file_type: string | null;
  storage_path: string;
  status: "uploaded" | "processing" | "ready" | "failed";
  created_at: string;
};

type UploadDocumentResponse = {
  message: string;
  bucket: string;
  storage_path: string;
  document: ApiDocument;
};

type ProcessDocumentResponse = {
  message: string;
  data: {
    document: ApiDocument;
    chunks_created: number;
    status: "ready";
  };
};

export async function uploadDocument(params: {
  subjectId: string;
  file: File;
}): Promise<ApiDocument> {
  const formData = new FormData();

  formData.append("subject_id", params.subjectId);
  formData.append("file", params.file);

  const response = await fetch(`${API_URL}/documents/upload`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(error?.detail ?? "Error al subir documento");
  }

  const result: UploadDocumentResponse = await response.json();

  return result.document;
}

export async function processDocument(
  documentId: string
): Promise<ProcessDocumentResponse["data"]> {
  const response = await fetch(`${API_URL}/documents/${documentId}/process`, {
    method: "POST",
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(error?.detail ?? "Error al procesar documento");
  }

  const result: ProcessDocumentResponse = await response.json();

  return result.data;
}